src/nvme/mi.c

```diff
@@ -611,9 +638,42 @@ int nvme_mi_admin_xfer(nvme_mi_ctrl_t ctrl,
 	resp.data_len = *resp_data_size;
 
 	/* limit the response size, specify offset */
-	admin_req->flags = 0x3;
-	admin_req->dlen = cpu_to_le32(resp.data_len & 0xffffffff);
-	admin_req->doff = cpu_to_le32(resp_data_offset & 0xffffffff);
+	if (direction == NVME_DATA_TFR_HOST_TO_CTRL)
+	{
+		if (req.data_len)
+		{
+			admin_req->flags = 0x1;
+		}
+		else
+		{
+			admin_req->flags = 0;
+		}
+		admin_req->dlen = cpu_to_le32(resp.data_len & 0xffffffff);
+		admin_req->doff = 0;
+	}
+	else if (direction == NVME_DATA_TFR_CTRL_TO_HOST)
+	{
+		if (resp.data_len && resp_data_offset)
+		{
+			admin_req->flags = 0x3;
+		}
+		else if (resp.data_len)
+		{
+			admin_req->flags = 0x1;
+		}
+		else
+		{
+			admin_req->flags = 0x0;
+		}
+		admin_req->dlen = cpu_to_le32(resp.data_len & 0xffffffff);
+		admin_req->doff = cpu_to_le32(resp_data_offset & 0xffffffff);
+	}
+	else
+	{
+		admin_req->flags = 0;
+		admin_req->dlen = 0;
+		admin_req->doff = 0;
+	}
```

Comment:
I'm not sure how much of this change is required; it's perfectly valid to have the flags at `0x3`, and the offsets still zero. In fact, the flags are deprecated in NVMe-MI 1.2, and the dlen/dofst values are always considered valid, so we'll need to do that anyway.